push/ups_integration_with_openshift.md

```diff
@@ -0,0 +1,144 @@
+
+*As a developer, I want to write custom logic for sending Push messages so that I can dynamically send messages based on varying factors and inputs*
+
+Push messages can be sent by calling the [sender API](https://aerogear.org/docs/specs/aerogear-unifiedpush-rest/index.html#397083935) with the correct key. There are features available in the OpenShift and the Catalog that can simplify the provisoining of a custom service, and obtaining secret credentials (like a key) from a provisioned service.
+
+## OAuth provider support 
+
+To address the first & second stories, UPS will be modified to support different auth providers.
+The 2 providers will be Keycloak (to maintain backwards compatibility for existing users) and a generic OAuth Provider.
+The OAuth flow will follow the same pattern & flow as [Prometheus](https://github.com/aerogearcatalog/metrics-apb/blob/9b4cb90988f5f3e7a28a84050cf274355cd36498/roles/provision-metrics-apb/tasks/provision-prometheus.yml#L36-L52) with the [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy).
+The UPS APB will be modified to include an OAuth ServiceAccount and the OAuth Proxy container.
+UPS will be modified to allow running in 'no auth' mode i.e. the UI & API will be uprotected and give 'admin' access to the UPS instance running in the user's namespace. The OAuth Proxy will sit in front of UPS providing authentication and autorisation based on namespace permissions.
+
+A future change to this could look at adding support for individual 'developer' users.
+This would be required if running UPS as a shared service.
+That would most likely require changes to UPS to call OpenShift to check user permissions based on the token header, and only showing that developers Push Applications.
+
+These changes will mean UPS will no longer have a dependency on Keycloak when running in OpenShift.
+Also, the developer doesn't have to know about or be concerned with where or how the auth provider is setup and configured for UPS.
+
+## OpenShift UI Integration
+
+To address the third story, a number of UI screens and integration points will be added to the OpenShift UI & the UPS APB.
+
+Here is an example workflow showing the UI touchpoints and the idea of 'linking' Variants to MobileClients in OpenShift:
+
+* developer creates a Cordova App (MobileClient) in OpenShift
+* developer provisions UPS from the Catalog
+* developer sets up the FCM/Android credentials for push for the Cordova App via *new* UI screens in OpenShift
+* this creates an Android variant in UPS (details of how this happens TBD) and links it to the Mobile Client (details of how this 'link' works TBD)
+* the Variant config is retrieved from UPS and stored in/with the Cordova App resource (MobileClient)
```

Comment:
This means we have two sources of configuration for variants. See this [comment](https://issues.jboss.org/browse/AEROGEAR-2326?focusedCommentId=13547700&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13547700) for why this is a problem.

Reply:
The 2 main options:

1) OpenShift as the source of truth:
* Syncing *From* OpenShift *to* UPS
  * New Secrets mean a Variant is created
  * Updates to Secrets overwrite the Variant
  * Deleting a Secret will delete the Variant
  * Deleting a Variant will cause it to be recreated
* Syncing *From* UPS *to* OpenShift
  * Variant config will be annotated on the Secret

2) UPS as the source of truth:
* Syncing *From* UPS *to* OpenShift
  * New Variants are *not* synced to OpenShift
  * Updates to Variants that are backed by a Secret get synced back to the Secret
  * Deleting a Variant that is backed by a Secret will also delete that Secret
  * Variant config will be annotated on the Secret
* Syncing *From* OpenShift *to* UPS
  * New Secrets mean a Variant is created
  * Updates to Secrets *don't* get updated in Variants
  * Deleting a Secret has no action in UPS (Variant remains)

Option 1 problems are:
* making changes in the UPS UI will be overwritten on next Sync

Option 2 problems are:
* updating the Secret with new FCM/Apple credentials don't get propagated to UPS



What about a 3rd option where changes are synced both ways?
The conflict resolution strategy could be kept simple where the last in wins.
This could cause confusion (and possibly errors if something is deleted) in edge cases where both the Variant and Secret are modified at the same time.
However, maybe that's OK.

@maleck13 thoughts?

Reply:
+1 on Openshift as the source of truth

Reply:
After call with @matzew & @pb82, and comparing to how jenkins manages the syncing between openshift and jenkins pipelines & builds, we decided to go with option 1 for simplicity at this time